src/Microsoft.AspNetCore.Routing/Matching/DfaMatcherBuilder.cs

```diff
@@ -370,20 +414,20 @@ internal Candidate[] CreateCandidates(IReadOnlyList<Endpoint> endpoints)
         // internal for tests
         internal Candidate CreateCandidate(Endpoint endpoint, int score)
         {
-            var assignments = new Dictionary<string, int>(StringComparer.OrdinalIgnoreCase);
-            var slots = new List<KeyValuePair<string, object>>();
-            var captures = new List<(string parameterName, int segmentIndex, int slotIndex)>();
-            (string parameterName, int segmentIndex, int slotIndex) catchAll = default;
+            _assignments.Clear();
```

Comment:
Do you want to clear these lists at all when **endpoint** is not a **RouteEndpoint**?
In that case you could create a **Candidate** instance with only **Array.Empty&lt;T&gt;** instances
No idea how common this is though.